### PR TITLE
Accept GatewayClasses for this controller

### LINKED ISSUE #41: Add support for registering supported GatewayClasses

### PR DESCRIPTION
**Pull request by david-martin**
Closes #41 

To verify, run `make test`.

Alternatively, run the controller locally with the `run` target,
then create the sample gatewayclass `kubectl apply -f ./config/samples/gatewayclass.yaml`

The gatewayclass status conditions should go from
```
  - lastTransitionTime: "1970-01-01T00:00:00Z"
    message: Waiting for controller
    reason: Waiting
    status: Unknown
    type: Accepted
```
to 
```
  - lastTransitionTime: "2023-02-03T13:45:58Z"
    message: Handled by kuadrant.io/mctc-gw-controller
    reason: Accepted
    status: "True"
    type: Accepted
```

Changes:
- Set any GatewayClasses with a controller name of 'kuadrant.io/mctc-gw-controller' to accepted
- Add an example GatewayClass `mctc-gw-istio-instance-per-cluster`
- Add the gateway api CRDs @v0.6.0 for easier testing with envtest. The local-setup and kustomize config has been updated to reference the same CRDs now checked in.
- Add integration tests for the gatewayclass controller for accepted and not accepted gatewayclasses

### CODE REVIEW

**Comment by david-martin** [comment]
@mikenairn @philbrookes Interested in any feedback on this approach to controller integration testing https://github.com/Kuadrant/multi-cluster-traffic-controller/pull/56/files#diff-021c17f8f25b24711c10c8b1a724278646cacd00854dc5378a644115cf668d79

File: pkg/controllers/gateway/gatewayclass_controller.go [hunk 1]
```diff
@@ -46,9 +57,67 @@ type GatewayClassReconciler struct {
...
+		len(previous.Status.Conditions) == 1 &&
+		previous.Status.Conditions[0].Type == string(gatewayv1beta1.GatewayClassConditionStatusAccepted) &&
+		previous.Status.Conditions[0].Status == "True" {
+		log.Info("GatewayClass already Accepted")
+		return ctrl.Result{}, nil
+	}
+
+	gatewayclass := previous.DeepCopy()
+	supportedClasses := getSupportedClasses()
+	if !contains(supportedClasses, previous.Name) {
```
Comment (sergioifg94) [inline pkg/controllers/gateway/gatewayclass_controller.go:88]:
```suggestion
	if !slice.ContainsString(supportedClasses, previous.Name) {
```

The [`slice`](https://github.com/Kuadrant/multi-cluster-traffic-controller/blob/main/pkg/_internal/slice/slice.go#L22) package already has a function we could use

**Review by sergioifg94 — COMMENTED** [review]
Just a couple of notes from looking at the code

File: pkg/controllers/gateway/gatewayclass_controller.go [hunk 2]
```diff
@@ -57,6 +126,15 @@ func (r *GatewayClassReconciler) Reconcile(ctx context.Context, req ctrl.Request
 func (r *GatewayClassReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		// Uncomment the following line adding a pointer to an instance of the controlled resource as an argument
-		For(&gateway.GatewayClass{}).
+		For(&gatewayv1beta1.GatewayClass{}).
 		Complete(r)
```
[the remark this answers is not in the document]
Comment (david-martin) [reply pkg/controllers/gateway/gatewayclass_controller.go:130]: Nice. I was looking for an example of how to do this.
I'll test it out

File: pkg/controllers/gateway/gatewayclass_controller.go (hunk 1 above)
Comment (david-martin) [reply pkg/controllers/gateway/gatewayclass_controller.go:88]: Handy

File: pkg/controllers/gateway/gatewayclass_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sergioifg94) [reply pkg/controllers/gateway/gatewayclass_controller.go]: You need to import this package btw: `sigs.k8s.io/controller-runtime/pkg/predicate`

**Comment by david-martin** [comment]
Closing in favour of #59, which includes these changes as well

**Pull request closed without merging**